packages/contracts/src/contracts/current/forwarder/MixinBuyExactAssets.sol

```diff
@@ -0,0 +1,155 @@
+pragma solidity ^0.4.24;
+pragma experimental ABIEncoderV2;
+
+import "./MixinForwarderCore.sol";
+import "./MixinForwarderExpectedResults.sol";
+import "./MixinERC20.sol";
+import "./MixinERC721.sol";
+
+contract MixinBuyExactAssets is
+    MixinForwarderCore,
+    MixinForwarderExpectedResults,
+    MixinERC20,
+    MixinERC721
+{
+    uint8 public constant ERC20_PROXY_ID = 1;
+    uint8 public constant ERC721_PROXY_ID = 2;
+
+    /// @dev Buys the exact amount of assets (ERC20 and ERC721), performing fee abstraction if required.
+    ///      All order assets must be of the same type. Deducts a proportional fee to fee recipient.
+    ///      This function is payable and will convert all incoming ETH into WETH and perform the trade on behalf of the caller.
+    ///      The caller is sent all assets from the fill of orders. This function will revert unless the requested amount of assets are purchased.
+    ///      Any excess ETH sent will be returned to the caller
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param signatures An array of Proof that order has been created by maker.
+    /// @param feeOrders An array of Order struct containing order specifications for fees.
+    /// @param assetAmount The amount of maker asset to buy.
+    /// @param feeSignatures An array of Proof that order has been created by maker for the fee orders.
+    /// @param feeProportion A proportion deducted off the ETH spent and sent to feeRecipient. The maximum value for this
+    ///        is 1000, aka 10%. Supports up to 2 decimal places. I.e 0.59% is 59.
+    /// @param feeRecipient An address of the fee recipient whom receives feeProportion of ETH.
+    /// @return FillResults amounts filled and fees paid by maker and taker.
+    function buyExactAssets(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures,
+        uint256 assetAmount,
+        uint16  feeProportion,
+        address feeRecipient
+    )
+        payable
+        public
+        returns (Exchange.FillResults memory totalFillResults)
+    {
+        require(msg.value > 0, VALUE_GREATER_THAN_ZERO);
+        address takerAsset = readAddress(orders[0].takerAssetData, 0);
```

Comment:
This is doable, but I think I'm running into some popping issues. Where the length is modified. Since the string is not a compile time constant this feels like a problem.